src/sensor.c

```diff
@@ -0,0 +1,678 @@
+    if (contribution_rate == NULL)
+    {
+        /*
+        * Pointer Memory freed
+        */
+        free(sensor_array);
+        free(sd_matrix);
+
+        free(eigen->eigen_value);
+        for (int i = 0; i < no_of_sensors; i++)
+        {
+            free(eigen->eigen_vector[i]);
+        }
+
+        free(eigen->eigen_value);
+        free(eigen);
+        return INVALID_SENSOR_FUSION_VALUE;
+    }
+
+    /*
+    * Step 4 - determine_contribution_rates_to_use
+    */
+    int contribution_rates_to_use = determine_contribution_rates_to_use(contribution_rate, ((float)principal_component_ratio / 100.0), no_of_sensors);
+    if (contribution_rates_to_use <= 0)
+    {
+        /*
+        * Pointer Memory freed
+        */
+        free(sensor_array);
+        free(sd_matrix);
+
+        free(eigen->eigen_value);
+        for (int i = 0; i < no_of_sensors; i++)
+        {
+            free(eigen->eigen_vector[i]);
+        }
+
+        free(eigen);
+
+        free(contribution_rate);
+
+        return INVALID_CONTRIBUTION_RATES;
+    }
+
+    /*
+    * Step 5 - calculate_principal_components
+    */
+    double **principal_components_matrix = calculate_principal_components(sd_matrix, no_of_sensors, eigen->eigen_vector, contribution_rates_to_use);
+    if (principal_components_matrix == NULL)
+    {
+        /*
+        * Pointer Memory freed
+        */
+        free(sensor_array);
+        free(sd_matrix);
+
+        free(eigen->eigen_value);
+        for (int i = 0; i < no_of_sensors; i++)
+        {
+            free(eigen->eigen_vector[i]);
+        }
+
+        free(eigen->eigen_value);
+        free(eigen);
+
+        free(contribution_rate);
+        return INVALID_SENSOR_FUSION_VALUE;
+    }
+
+    /*
+    * Step 6 - calculate_integrated_support_degree_matrix
+    */
+    double *integrated_support_matrix =
+        calculate_integrated_support_degree_matrix(
+            principal_components_matrix,
+            contribution_rate,
+            contribution_rates_to_use, no_of_sensors);
+
+    if (integrated_support_matrix == NULL)
+    {
+        /*
+        * Pointer Memory freed
+        */
+        free(sensor_array);
+        free(sd_matrix);
+
+        free(eigen->eigen_value);
+        for (int i = 0; i < no_of_sensors; i++)
+        {
+            free(eigen->eigen_vector[i]);
+        }
+
+        free(eigen->eigen_value);
+        free(eigen);
+
+        free(contribution_rate);
+
+        for (int i = 0; i < no_of_sensors; i++)
+        {
+            free(principal_components_matrix[i]);
+        }
+        free(principal_components_matrix);
+
+        return INVALID_SENSOR_FUSION_VALUE;
+    }
+
+    /*
+    * Step 7 - eliminate_incorrect_data
+    */
+    int result_eliminate = eliminate_incorrect_data(integrated_support_matrix,
+                                                    ((float)q_support_value / 100.0), no_of_sensors);
+
+    if (result_eliminate < 0)
```

Comment:
You should start the curly bracket at the end of the first line.

Reply:
Changes have been done